worker: Restart streaming queries on a different tablet if the first …

michael-berlin opened the pull request:
…retry failed.

This is necessary when a tablet is temporary taken out of service.

I've added two integration tests in split_clone_test.go to cover the new functionality.

Other changes:
- Added stat variables for started and errored streaming queries.
- Fake query service uses the alias instead of the UID to make it easier to access the new stat vars by alias.
- The fake query service now supports an error callback to run code before an error is returned.
- TestSplitCloneV2_Offline_RestartStreamingQuery: Tightened the test by using only one source tablet and verifying the stat vars.
- Unified the clone code to always use a common prefix "table=%v chunk=%v" within the chunk pipeline.

@alainjobart

alainjobart commented on go/vt/worker/tablet_provider.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func (p *shardTabletProvider) getTablet() (*topodatapb.Tablet, error) {
+	// Pick any healthy serving source tablet.
```
Please add a comment here on why you do not use p.tsc.GetHealthyTabletStats (instead of RemoveUnhealthyTablets + GetTabletStats).

(and I am not sure why, by the way... in other places in the resharding code, we do this sometimes because we want to access a table even though its query service is off because of TabletControl, but here?)

alainjobart commented:
LGTM after comment is addressed. I like the TabletProvider interface. The definition of the 'retryable' flag could be a bit clearer though.

michael-berlin commented:
> The definition of the 'retryable' flag could be a bit clearer though.

I've added comments for the two methods which return a `retryable` boolean.

As discussed offline, I've also changed the code to use the list of healthy tablets from the tablet stats cache directly where it's suitable.
